stock_reserve_sale/models/sale.py

```diff
@@ -0,0 +1,233 @@
+from odoo import models, fields, api, _
+from odoo.exceptions import UserError
+from odoo.tools.float_utils import float_compare
+import odoo.addons.decimal_precision as dp
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    @api.multi
+    @api.depends('state',
+                 'order_line.reservation_ids',
+                 'order_line.is_stock_reservable')
+    def _compute_stock_reservation(self):
+        for sale in self:
+            has_stock_reservation = False
+            is_stock_reservable = False
+            for line in sale.order_line:
+                if line.reservation_ids:
+                    has_stock_reservation = True
+                if line.is_stock_reservable:
+                    is_stock_reservable = True
+            if sale.state not in ('draft', 'sent'):
+                is_stock_reservable = False
+            sale.is_stock_reservable = is_stock_reservable
+            sale.has_stock_reservation = has_stock_reservation
+
+    has_stock_reservation = fields.Boolean(
+        compute='_compute_stock_reservation',
+        readonly=True,
+        store=True,
+        string='Has Stock Reservations')
+    is_stock_reservable = fields.Boolean(
+        compute='_compute_stock_reservation',
+        readonly=True,
+        store=True,
+        string='Can Have Stock Reservations')
+
+    @api.multi
+    def release_all_stock_reservation(self):
+        lines = self.mapped('order_line')
+        lines.release_stock_reservation()
+        return True
+
+    @api.multi
+    def action_confirm(self):
+        self.release_all_stock_reservation()
+        return super(SaleOrder, self).action_confirm()
+
+    @api.multi
+    def action_cancel(self):
+        self.release_all_stock_reservation()
+        return super(SaleOrder, self).action_cancel()
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    @api.multi
+    def _get_line_rule(self):
+        self.ensure_one()
+        if self.route_id and self.route_id.rule_ids:
+            return self.route_id.rule_ids.sorted(
+                key=lambda r: (r.route_sequence, r.sequence))[0]
```

Comment:
Prefer using 'first()'
```suggestion
                from odoo.fields import first
                return first(self.route_id.rule_ids.sorted(
                key=lambda r: (r.route_sequence, r.sequence)))
```